### PR TITLE
retry policy for reads, and speculative execution, #168

### PR DESCRIPTION
**Pull request by patriknw**
Refs #168

### CODE REVIEW

File: src/main/resources/reference.conf [hunk 1]
```diff
@@ -57,6 +57,20 @@ cassandra-journal {
...
+  
+  # The number of retries when a read query fails.
+  read-retries = 3
+  
+  # Set this to a positive integer to enable speculative executions.
+  # The value defines the number of speculative executions that will be
+  # performed with the delay defined by 'speculative-executions-delay'.
+  # See http://docs.datastax.com/en/developer/java-driver/3.1/manual/speculative_execution/
+  # If you enable this you should also enable cassandra-query-journal.speculative-executions
+  # which is used for the relay query.
```
Comment (patriknw) [inline src/main/resources/reference.conf]: yeah, those duplicate configs are not nice, eventually we will change the config structure according to https://github.com/akka/akka-persistence-cassandra/issues/81

**Comment by patriknw** [comment]
For the record, I have used the following test and [CCM](https://github.com/pcmanus/ccm) to semi-manually test this.

```scala
package akka.persistence.cassandra.journal

import scala.concurrent.duration._
import akka.actor._
import akka.persistence._
import akka.testkit._
import com.typesafe.config.ConfigFactory
import org.scalatest._
import akka.persistence.cassandra.CassandraLifecycle

object CassandraIntegrationSpec2 {
  val config = ConfigFactory.parseString(
    s"""
      cassandra-journal.keyspace=CassandraIntegrationSpec
      cassandra-snapshot-store.keyspace=CassandraIntegrationSpecSnapshot
      cassandra-journal {
        log-queries = on

        replication-factor = 3
        contact-points = ["127.0.0.1", "127.0.0.2", "127.0.0.3"]

        #local-datacenter = "datacenter1"
        #data-center-replication-factors = ["datacenter1:3","datacenter2:3"]
        #replication-strategy = "NetworkTopologyStrategy"

        port=9042
        write-consistency = "QUORUM"
        read-consistency = "QUORUM"

        speculative-executions = 1
      }

      cassandra-snapshot-store {
        log-queries = on
        replication-factor = 3
        contact-points = ["127.0.0.1", "127.0.0.2", "127.0.0.3"]

        #local-datacenter = "datacenter1"
        #data-center-replication-factors = ["datacenter1:3"]
        #data-center-replication-factors = ["datacenter1:3","datacenter2:3"]
        #replication-strategy = "NetworkTopologyStrategy"

        port=9042
        write-consistency = "QUORUM"
        read-consistency = "QUORUM"

        speculative-executions = 1
      }

      cassandra-query-journal.speculative-executions = 1
    """
  ).withFallback(CassandraLifecycle.config)

  class TestBot(override val persistenceId: String, probe: ActorRef) extends PersistentActor with ActorLogging {

    import context.dispatcher
    var state: Int = 0
    var time = System.nanoTime()

    private def scheduleAdd(): Unit =
      context.system.scheduler.scheduleOnce(1.millis, self, "add")

    override protected def onPersistFailure(cause: Throwable, event: Any, seqNr: Long): Unit = {
      super.onPersistFailure(cause, event, seqNr)
      probe ! cause.getMessage
    }

    override protected def onRecoveryFailure(cause: Throwable, event: Option[Any]): Unit = {
      super.onRecoveryFailure(cause, event)
      probe ! cause.getMessage
    }

    def receiveRecover: Receive = {
      case SnapshotOffer(_, snap: Int) =>
        state = snap
      case "add" =>
        //        log.info("# recover add => {}", state + 1)
        state += 1
      case RecoveryCompleted =>
        val duration = (System.nanoTime() - time).nanos
        log.info("# RecoveryCompleted in {} ms => {}", duration.toMillis, state + 1)
        if (duration > 15.seconds)
          probe ! s"recovery took ${duration.toMillis} ms"
        time = System.nanoTime()
        scheduleAdd()
    }

    def receiveCommand: Receive = {
      case "add" =>
        log.info("# command add => {}", state)
        time = System.nanoTime()
        persist("add") { _ =>
          val duration = (System.nanoTime() - time).nanos
          log.info("# event add in {} ms => {}", duration.toMillis, state + 1)
          if (duration > 5.seconds)
            probe ! s"persist ${state + 1} took ${duration.toMillis} ms"
          state += 1
          if (state % 1000 == 0) {
            log.info("# stopping => {}", state)
            probe ! "normal stop"
            context.stop(self)
          } else {
            if (state % 10000 == 0)
              saveSnapshot(state)
            scheduleAdd()
          }
        }
    }

  }

}

// CCM is useful: https://github.com/pcmanus/ccm
class CassandraIntegrationSpec2 extends TestKit(ActorSystem("CassandraIntegrationSpec2", CassandraIntegrationSpec2.config))
  with ImplicitSender with WordSpecLike with Matchers {

  import CassandraIntegrationSpec2._

  "A Cassandra journal" should {
    "survive Cassandra server failures" in {
      val persistenceId = "p1"
      (1 to 100).foreach { n =>
        println(s"# RESTARTING $n") // FIXME
        Thread.sleep(5000)
        val ref1 = system.actorOf(Props(classOf[TestBot], persistenceId, testActor))
        watch(ref1)
        within(60.seconds) {
          expectMsg("normal stop")
          expectTerminated(ref1)
        }
      }
    }
  }

}
```

File: src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala [hunk 2]
```diff
@@ -542,10 +545,27 @@ private[cassandra] object CassandraJournal {
...
-  override def onUnavailable(statement: Statement, cl: ConsistencyLevel, requiredReplica: Int, aliveReplica: Int, nbRetry: Int): RetryDecision = retry(cl, nbRetry)
-  override def onWriteTimeout(statement: Statement, cl: ConsistencyLevel, writeType: WriteType, requiredAcks: Int, receivedAcks: Int, nbRetry: Int): RetryDecision = retry(cl, nbRetry)
-  override def onReadTimeout(statement: Statement, cl: ConsistencyLevel, requiredResponses: Int, receivedResponses: Int, dataRetrieved: Boolean, nbRetry: Int): RetryDecision = retry(cl, nbRetry)
-  override def onRequestError(statement: Statement, cl: ConsistencyLevel, cause: DriverException, nbRetry: Int): RetryDecision = tryNextHost(cl, nbRetry)
+  override def onUnavailable(statement: Statement, cl: ConsistencyLevel, requiredReplica: Int, aliveReplica: Int,
+                             nbRetry: Int): RetryDecision = {
+    if (nbRetry == 0)
+      tryNextHost(cl, nbRetry) // see DefaultRetryPolicy
+    else
+      retry(cl, nbRetry)
```
Comment (johanandren) [inline src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala]: So, go to next host if this is the first failure, but then stay with the next host? What if the next host also fails?

File: src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala [hunk 3]
```diff
@@ -542,10 +545,27 @@ private[cassandra] object CassandraJournal {
 }
 
 class FixedRetryPolicy(number: Int) extends RetryPolicy {
-  override def onUnavailable(statement: Statement, cl: ConsistencyLevel, requiredReplica: Int, aliveReplica: Int, nbRetry: Int): RetryDecision = retry(cl, nbRetry)
-  override def onWriteTimeout(statement: Statement, cl: ConsistencyLevel, writeType: WriteType, requiredAcks: Int, receivedAcks: Int, nbRetry: Int): RetryDecision = retry(cl, nbRetry)
-  override def onReadTimeout(statement: Statement, cl: ConsistencyLevel, requiredResponses: Int, receivedResponses: Int, dataRetrieved: Boolean, nbRetry: Int): RetryDecision = retry(cl, nbRetry)
-  override def onRequestError(statement: Statement, cl: ConsistencyLevel, cause: DriverException, nbRetry: Int): RetryDecision = tryNextHost(cl, nbRetry)
+  override def onUnavailable(statement: Statement, cl: ConsistencyLevel, requiredReplica: Int, aliveReplica: Int,
```
Comment (johanandren) [inline src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala]: A little explanatory scaladoc on this policy class wouldn't hurt.

Comment (patriknw) [inline src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala]: Added, it's essentially explained in http://docs.datastax.com/en/developer/java-driver/3.1/manual/retries/

File: src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala (hunk 2 above)
Comment (patriknw) [inline src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala]: Implementation stolen from DefaultRetryPolicy. I have added same description as a comment.

File: src/main/resources/reference.conf [hunk 4]
```diff
@@ -57,6 +57,20 @@ cassandra-journal {
...
+  speculative-executions = 0
```
Comment (ktoso) [inline src/main/resources/reference.conf]: nice nice!

**Pull request merged**